# What is interning?

- Let's revisit a past example:

In [1]:
a = 10
b = 10
id(a), id(b)

(1990427024, 1990427024)

- `a` and `b` are both integers with value 10
    - To save memory, Python has assigned them to the same memory address
        - **Recall**: this should be fine since integers are immutable objects

In [2]:
a = 500
b = 500
id(a), id(b)

(2705010545648, 2705010545520)

- In this case, however, `a` and `b` **don't have the same memory address**
    - *What's going on?*
        - This is Python **interning**

- When Python (more specifically CPython) pre-loads the integers between -5 and 256
    - When we reference an integer in that range, instead of creating a new version of it at a different memory address, Python will just reference the pre-loaded one

- Essentially, the integers in the range [-5, 256] are **singletons**
    - Only instantiated once
    
- *Why does Python bother pre-loading these integers?*
    - Small integers show up ofter in most code
        - Worth it to try to save some time

- Let's show that what we're asserting is actually true:

In [3]:
i = -5

while True:
    str_code_a = f"a = {i}"
    str_code_b = f"b = {i}"
    exec(str_code_a)
    exec(str_code_b)
    if a is not b:
        break
    i += 1
print(i)

257


- As we can see, the first integer where this isn't true is 257